### ISSUE TITLE
CoreDNS resolution failure for external hostnames with "overflow unpacking uint32"

### ISSUE THREAD

**Issue by fabn** [coredns/coredns#3305]
First of all forgive me if this is not the right place to post. I'm using coredns (in a k8s/rancher installation). Rancher uses its own fork of coredns, but looking at their repository it seems to be the same code of coredns itself.

Here's my original issue: rancher/rke/issues/1662 but I think that I've posted it in the wrong place.

Their rke tool install CoreDNS-1.3.1 with this configuration

```
.:53 {
    errors
    health
    kubernetes cluster.local in-addr.arpa ip6.arpa {
      pods insecure
      upstream
      fallthrough in-addr.arpa ip6.arpa
      ttl 30
    }
    prometheus :9153
    forward . '/etc/resolv.conf'
    cache 30
    loop
    reload
    loadbalance
}
```

Log start:

```
.:53
2019-09-24T13:54:37.187Z [INFO] CoreDNS-1.3.1
2019-09-24T13:54:37.187Z [INFO] linux/amd64, go1.11.4, 6b56a9c
CoreDNS-1.3.1
linux/amd64, go1.11.4, 6b56a9c
```

I've installed a rancher cluster using RKE with 3 nodes on KVM, until now everything worked well, but starting today I have an (intermittent) issue in DNS names resolution in my pods. It happens only for some hosts and randomly.

When it happens dns resolution (for external names) in pods is not working and I get some nasty errors in coredns pod, here are some examples:

```
2019-09-24T12:46:25.111Z [INFO] plugin/reload: Running configuration MD5 = 45cd9f91917cc54711e243e0d08537a7
2019-09-24T12:47:27.474Z [ERROR] plugin/errors: 2 security.ubuntu.com. A: dns: overflow unpacking uint32
2019-09-24T12:47:32.475Z [ERROR] plugin/errors: 2 security.ubuntu.com. A: dns: overflow unpacking uint32
2019-09-24T12:47:37.476Z [ERROR] plugin/errors: 2 security.ubuntu.com. A: dns: overflow unpacking uint32
2019-09-24T13:12:39.537Z [ERROR] plugin/errors: 2 registry.npmjs.org. A: dns: overflow unpacking uint32
2019-09-24T13:12:39.549Z [ERROR] plugin/errors: 2 registry.npmjs.org. AAAA: dns: overflow unpacking uint16
2019-09-24T13:12:44.539Z [ERROR] plugin/errors: 2 registry.npmjs.org. AAAA: dns: overflow unpacking uint16
2019-09-24T13:12:44.543Z [ERROR] plugin/errors: 2 registry.npmjs.org. A: dns: overflow unpacking uint32   
```

I've enabled coredns logs directive in configmap to have verbose logging and I get this:

```
2019-09-24T13:13:56.246Z [INFO] 10.42.0.230:56169 - 21605 "A IN registry.npmjs.org. udp 36 false 512" SERVFAIL qr,rd 36 5.003233501s
2019-09-24T13:13:56.246Z [ERROR] plugin/errors: 0 registry.npmjs.org. A: dns: overflow unpacking uint32
2019-09-24T13:13:56.251Z [INFO] 10.42.0.230:56169 - 4205 "AAAA IN registry.npmjs.org. udp 36 false 512" SERVFAIL qr,rd 36 5.008031338s
2019-09-24T13:13:56.251Z [ERROR] plugin/errors: 0 registry.npmjs.org. AAAA: dns: overflow unpacking uint16
```

Finally I also changed this directive in `Corefile`
```
forward . "/etc/resolv.conf"
# replaced with
forward . 1.1.1.1
# also tried with
forward . 8.8.8.8
```

Nothing changes.

I can find nothing relevant by googling for message "overflow unpacking uint32", just some  code fragment where is triggered.

My coredns deplyoment uses rancher/coredns-coredns:1.3.1 as image.

What can be?

**Comment by miekg** [comment]
something is sending a packet that doesn't fit in the UDP buffer size. This used to behavior in miekg/dns but that was fixed a while back. The query should be retried with TCP, as hopefully the TC bit (truncated) is also set

**Comment by fabn** [comment]
Today issue disappeared for A records, but it's still present for AAAA records, so name resolution is working for A records. I've also updated coredns to 1.6.2 but issue persists.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] CoreDNS resol..." ]
>Today issue disappeared for A records, but it's still present for AAAA records, so name resolution is working for A records. I've also updated coredns to 1.6.2 but issue persists.

without tcpdump to catch one these in the act, this won't be solved.

**Comment by fabn** [comment]
Obviously, I'll try to grab some packets, any hint on command to use, and format of packet to save? Do I need to grab packets on container or in the host machine?

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] CoreDNS resol..." ]
>Obviously, I'll try to grab some packets, any hint on command to use, and format of packet to save? Do I need to grab packets on container or in the host machine?

tcpdump should suffice with various flags to show, tcpdump port 53 -w dump.pcap 
should suffice so I can load it in wireshark. Pod or host shouldn't matter.

**Comment by fabn** [comment]
The issue is back and this time isn't going away. I was able to capture packets when this happens, here's the link: https://gofile.io/?c=2CX96f

I also see #3367 that can be related.

Please advice if you can.

**Comment by miekg** [comment]
eyeballing that dump. I see you ask git.rancher.io with a upd buffer of 512 bytes (no EDNS0) and the reply comes back that is 599 bytes... that will never work.

We currently do
~~~
	// Set buffer size correctly for this client.
	pc.c.UDPSize = uint16(state.Size())
	if pc.c.UDPSize < 512 {
		pc.c.UDPSize = 512
	}
~~~
It *may* be prudent to just always do a query with a larger buffer and then cut is down for this client, this may also help for caching. 

Can you cut down on the number of plugins? The only important ones here are cache and forward, interested in running with just forward and then with forward+cache

**Comment by fabn** [comment]
> I see you ask git.rancher.io with a upd buffer of 512 bytes (no EDNS0) and the reply comes back that is 599 bytes... that will never work.

I understand the issue. However the client is docker itself. This is running inside a kubernetes cluster. Here's the coredns configuration of dns pod:

```
.:53 {
    errors
    health
    kubernetes cluster.local in-addr.arpa ip6.arpa {
      pods insecure
      upstream
      fallthrough in-addr.arpa ip6.arpa
      ttl 30
    }
    prometheus :9153
    forward . 8.8.8.8 1.1.1.1
    cache 30
    loop
    reload
    loadbalance
}
```

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] CoreDNS resol..." ]
>> I see you ask git.rancher.io with a upd buffer of 512 bytes (no EDNS0) and the reply comes back that is 599 bytes... that will never work.
>I understand the issue. However the client is docker itself. This is running inside a kubernetes cluster. Here's the coredns configuration of dns pod:

docker itself, you mean the Go code in docker? Then they are doing it (somewhat) 
wrong. The upstream sending back this response is *not* compliant with the DNS 
standard. It should not send back such a response.

You can maybe work around it by setting force_tcp option in the forward stanza.

tl;dr: various pieces seem to violate the DNS standard. I *think* coredns is 
doing the right thing here and I'm reluctant to add (non-standard) work arounds

**Comment by fabn** [comment]
> docker itself, you mean the Go code in docker? Then they are doing it (somewhat) wrong. 

Yes, it should be go code in docker. I was able to reproduce the issue outside kubernetes and coredns.

I get a similar issue by using a plain `docker pull hello-world` command in the same machine where coredns is running and **also** in another machine in the same network. Here's the output:

```
docker pull hello-world
Error while pulling image: Get https://index.docker.io/v1/repositories/library/hello-world/images: dial tcp: lookup index.docker.io on 8.8.4.4:53: cannot unmarshal DNS message
```

If I do a `dig/curl/nslookup` for index.docker.io everything works, so it must be something inside the go code that causes this issue.

> The upstream sending back this response is *not* compliant with the DNS standard. It should not send back such a response. 

This is very unlikely, I changed upstream servers with both `8.8.8.8` (google) and `1.1.1.1` (cloudflare) and the issue persists. Also is strange that Google's dns sends a malformed response.

I also tried another approach: I installed dnsmasq locally using `8.8.8.8` as upstream server, then I've updated local `/etc/resolv.conf` with `nameserver 127.0.0.1` and everything worked, I'm able to pull images from registries and docker works.

I'm starting to think that the issue could be in the LAN itself, main gateway (from Microtik) which runs RouterOS could have some issues with EDNS. I found these resources (even if they are pretty old) on the web, so I'll point my investigation in that direction.

https://www.dns-oarc.net/oarc/services/replysizetest
https://forum.mikrotik.com/viewtopic.php?t=46227

> You can maybe work around it by setting force_tcp option in the forward stanza. tl;dr: various pieces seem to violate the DNS standard. I *think* coredns is doing the right thing here and I'm reluctant to add (non-standard) work arounds

My attempts clearly shows that coredns is not involved in this issue, so I think this could be closed. I'll try to use `force_tcp` as temporary workaround until I don't find the actual issue with my setup.

Thanks for your support any way.

P.S. in order to continue my investigations, since I'm not so expert in DNS protocol could you please tell me how can I distinguish EDNS requests and responses in PCAP dumps?

**Comment by fabn** [comment]
Another question useful for my investigation, you said 

> The upstream sending back this response is not compliant with the DNS standard

I want to try to dump the same kind of packets by excluding the mikrotik router from the chain. How can I compare that with my previous dump? How can I say if a response is malformed or not?

Thanks again.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] CoreDNS resol..." ]
>Another question useful for my investigation, you said
>
>> The upstream sending back this response is not compliant with the DNS standard
>
>I want to try to dump the same kind of packets by excluding the mikrotik router from the chain. How can I compare that with my previous dump? How can I say if a response is malformed or not?

Haven't looked. But a request without an additonal section, defaults to a udp 
buffer of 512 bytes, if the one-the-wire response exceeds that you get the error 
you see. This is bad and the *upstream*'s fault - it should never do that.

Re: docker code: that's also Go code and also uses miekg/dns as the lib (I 
think), although I can also use: net-go's resolver (which is pretty sane), or 
libc (all bets are off in that cause). But this doesn't matter that much, the 
upstream is sending a packet that is too large (> 512B), that *seems* the root 
cause of this issue. What's the upstream? route53?

**Comment by fabn** [comment]
> What's the upstream? route53?

Currently both `8.8.8.8` and `1.1.1.1` this is leading me to think that the issue is in the LAN gateway. It's very unlikely that two different upstreams from different vendors causes the same malformed response.

**Comment by doubled1** [comment]
Commenting in case somebody else needs it.  I'm also using a Mikrotik router.

I noticed this while troubleshooting Nextcloud News.  There's a scheduled run that makes a request to a bunch of URLs for RSS updates, and some but not all were failing.  Looking at the CoreDNS logs gave many similar entries:

```
[ERROR] plugin/errors: 2 security.archlinux.org. AAAA: dns: overflow unpacking uint16                                       
[ERROR] plugin/errors: 2 opsgenie.status.atlassian.com. A: dns: overflow unpacking uint16                                   
```

On my Mikrotik `Max UDP Packet Size` defaults to 4096.  I've reduced mine to 512.  No errors in logs, and feeds have updated successfully.

**Comment by chief93** [comment]
Had simillar issue, but another message:
```
[ERROR] plugin/errors: 2 production.cloudflare.docker.com. A: dns: overflowing header size
```
Increasing the default value of `4096` to `8192` max UDP packet size in my `Mikrotik` seems solved this

**Comment by TheIdhem** [comment]
> Had simillar issue, but another message:
> 
> ```
> [ERROR] plugin/errors: 2 production.cloudflare.docker.com. A: dns: overflowing header size
> ```
> 
> Increasing the default value of `4096` to `8192` max UDP packet size in my `Mikrotik` seems solved this

I think that this answer is not correct because the valid range for max UDP packet size is between 512 to 4096. according to this link: https://coredns.io/plugins/bufsize/

**Comment by aa294828045** [comment]
在我的 Mikrotik 上Max UDP Packet Size默认为 4096。我已将我的减少到 512。日志中没有错误，并且提要已成功更新。

![image](https://github.com/coredns/coredns/assets/31462133/ffea96ee-7c35-49ea-872b-bb3dd0849c53)

**Comment by DenisBessa** [comment]
> 在我的 Mikrotik 上Max UDP Packet Size默认为 4096。我已将我的减少到 512。日志中没有错误，并且提要已成功更新。
> 
> ![image](https://user-images.githubusercontent.com/31462133/245978612-ffea96ee-7c35-49ea-872b-bb3dd0849c53.png)

Reducing the Max UDP packet size on my router solved the problem.

**Comment by fuog** [comment]
i did run in to the same problem with my homelab setup.

- k3s kubernetes, used clusterDNS `rancher/mirrored-coredns-coredns:1.10.1`
- requested DNS record : `resources.download.minecraft.net`
- error on coreDNS : `[ERROR] plugin/errors: 2 resources.download.minecraft.net. A: dns: overflow unpacking uint16
`
- or for example this: `[ERROR] plugin/errors: 2 minecraft.azureedge.net. AAAA: dns: overflow unpacking uint32
`

 i just wanted run a minecraft server that requested some resources via api from azure. the requested dns record on a normal Linux client  (to my Mikrotik Cloud Core Router) had no problem. a pod inside kubernetes did not get that dns resolution. (query is just stuck)

actually limiting my Mikrotik to 512 helped me but all it did was switching from UDP to TCP on both, pods in the cluster and any regular linux systems that had no issues before.

Regular Linux to mikrotik DNS working normal
 ```
❯ nslookup resources.download.minecraft.net 172.20.20.1
Server:		172.20.20.1
Address:	172.20.20.1#53

Non-authoritative answer:
resources.download.minecraft.net	canonical name = mojang-resourcesdownloadminecra.azureedge.net.
mojang-resourcesdownloadminecra.azureedge.net	canonical name = mojang-resourcesdownloadminecra.afd.azureedge.net.
mojang-resourcesdownloadminecra.afd.azureedge.net	canonical name = star-azureedge-prod.trafficmanager.net.
star-azureedge-prod.trafficmanager.net	canonical name = dual.part-0032.t-0009.t-msedge.net.
dual.part-0032.t-0009.t-msedge.net	canonical name = part-0032.t-0009.t-msedge.net.
Name:	part-0032.t-0009.t-msedge.net
Address: 13.107.246.60
Name:	part-0032.t-0009.t-msedge.net
Address: 13.107.213.60
Name:	part-0032.t-0009.t-msedge.net
Address: 2620:1ec:46::60
Name:	part-0032.t-0009.t-msedge.net
Address: 2620:1ec:bdf::60
```

Here the pod after switching to a max of 512 the client moves on to TCP-Mode that i never needed before.
```
 debugging-pod  ~  nslookup resources.download.minecraft.net
;; Truncated, retrying in TCP mode.    <--------------------------
Server:		100.65.0.10
Address:	100.65.0.10#53

Non-authoritative answer:
resources.download.minecraft.net	canonical name = mojang-resourcesdownloadminecra.azureedge.net.
..... same ..
```
my question would be: why sould my entire network adapt to that limit when only coreDNS can't handle that UDP package size ?. I did that but i don't see it realy. :man_shrugging: 

I hope this helps someone else too

**Comment by chrisohaver** [comment]
> why sould my entire network adapt to that limit when only coreDNS can't handle that UDP package size ?. I did that but i don't see it realy.

Commentary above seems to suggest that this is a problem with an upstream DNS server creating an illegal overflowed response, and upon receiving these malformed responses, CoreDNS _justifiably_ rejects them.

Would it be nice if CoreDNS could work around these upstream bugs?  Yes, there is [workaround patch](https://github.com/coredns/coredns/pull/6277) that aims to fix this sort of issue already merged and to be included in the next release (a date for the next has not yet been planned).

**Comment by NfNitLoop** [comment]
> a date for the next has not yet been planned

We recently started bumping into this problem, in particular when querying for AWS S3 hostnames in the format `<bucketName>.s3.us-west-1.amazonaws.com`. I assume the reason is that we get back an answer with the CNAME and also a lot of A records for the CNAME, overflowing the buffer. Glad to hear a patch will be released sometime!

In the meantime, (in case anyone else is arriving at this page from web searches as I did), we worked around it by configuring our k8s pods to prefer TCP DNS resolution. You just need to add the [use-vc] option to your [dnsConfig].options like this:

```yaml
spec:
  # …
  dnsConfig:
    options:
      - name: use-vc
```
[use-vc]: https://man7.org/linux/man-pages/man5/resolv.conf.5.html
[dnsConfig]: https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#pod-dns-config

**Comment by stdll00** [comment]
In docker-desktop kubernetes cluster (coredns 1.10.1 ) , I cannot resolve `oauth2.googleapis.com` from k8s pod. Coredns side error is  `oauth2.googleapis.com` because `[ERROR] plugin/errors: 2 oauth2.googleapis.com. A: dns: overflow unpacking uint32.
I changed coredns image version to 1.10.0 (registry.k8s.io/coredns/coredns:v1.10.0) , I can resolve `oauth2.googleapis.com`.  Not works with `1.11.1`
So I guess some bug exists between 1.10.0 and 1.10.1

**Comment by tannertaylor** [comment]
> In docker-desktop kubernetes cluster (coredns 1.10.1 ) , I cannot resolve `oauth2.googleapis.com` from k8s pod. Coredns side error is `oauth2.googleapis.com` because `[ERROR] plugin/errors: 2 oauth2.googleapis.com. A: dns: overflow unpacking uint32. I changed coredns image version to 1.10.0 (registry.k8s.io/coredns/coredns:v1.10.0) , I can resolve `oauth2.googleapis.com`. Not works with `1.11.1` So I guess some bug exists between 1.10.0 and 1.10.1

That fixed it for me as well. I was also just using Docker Desktop but would get that same error for both oauth2.googleapis.com and login.microsoftonline.com. Reverting the coredns deployment image to 1.10.0 fixed it for me.

**Comment by aloutfi** [comment]
This issue persists in the most recent version of CoreDNS (v1.11.1), specifically affecting login.microsoftonline.com. It has become apparent in scenarios where developers, unaware of the underlying compatibility issues, update their [Docker Desktop to versions beyond 4.21.0](https://docs.docker.com/desktop/release-notes/#4210). This particular Docker Desktop update included an upgrade to CoreDNS v1.10.1. Subsequent versions of Docker Desktop have incorporated CoreDNS newer than this, indicating that this problem now constitutes a regression affecting the last three releases of CoreDNS.


Following the guidance of @stdll00 and @tannertaylor, I opted to downgrade to CoreDNS v1.10.0 using the following commands:

```bash
kubectl edit deployment coredns -n kube-system
# Modify the image to: image: registry.k8s.io/coredns/coredns:v1.10.0 (around line 52)
kubectl rollout status deployment/coredns -n kube-system
```
Once rollout is complete I am able to hit login.microsoftonline.com.

**Comment by steven-so** [comment]
I was running into 
```
[ERROR] plugin/errors: 2 tortuga-prod-na.s3-external-1.amazonaws.com. A: dns: overflow unpacking uint16
```
Luckily I found this, and downgrading to CoreDNS v1.10.0 fixed my problem. Thanks!

**Comment by amartincolby** [comment]
I'm running Microk8s and encountered similar problems. I was forced to downgrade to a year-old version of Microk8s to go from 1.10.1 to 1.10.0.

**Comment by berginv** [comment]
```yaml
spec:
  # …
  dnsConfig:
    options:
      - name: edns0
```

This fixed my issue 👍🏿
